software/bsg_manycore_lib/bsg_striped_array.hpp

```diff
@@ -0,0 +1,69 @@
+#pragma once
+extern "C" {
+#include "bsg_manycore.h"
+}
+
+namespace bsg_manycore {
+    template <typename TYPE, int SIZE, int TG_SIZE_X, int TG_SIZE_Y, int STRIPE_SIZE=1>
+    class TileGroupStripedArray {
+    public:
+        static constexpr int TILES = TG_SIZE_X * TG_SIZE_Y;
+        static constexpr int STRIPES = SIZE/STRIPE_SIZE + SIZE%STRIPE_SIZE;
+        static constexpr int STRIPES_PER_TILE = (STRIPES/TILES) + (STRIPES%TILES == 0 ? 0 : 1);
+        static constexpr int ELEMENTS_PER_TILE = STRIPES_PER_TILE * STRIPE_SIZE;
+
+        TileGroupStripedArray() {}
+
+        int size() const { return SIZE; }
+        int stripe_size() const { return STRIPE_SIZE; }
+
+        TYPE operator[](int i) const {
+            return *_address(i);
+
+        }
+
+        TYPE & operator[](int i) {
+            return *_address(i);
+        }
+
+        TYPE at_local(int i) const {
+            return _data[i];
+        }
+
+        TYPE & at_local(int i) {
+            return _data[i];
+        }
+
+    private:
+        int _stripe(int i) const {
+            return i / STRIPE_SIZE;
+        }
+
+        int _word_in_stripe(int i) const {
+            return i % STRIPE_SIZE;
+        }
+
+        int _tile(int stripe) const {
+            return stripe % TILES;
+        }
+
+        int _stripe_in_tile(int stripe) const {
+            return stripe / TILES;
+        }
+
+        TYPE * _address(int i) const {
+            int stripe = _stripe(i);
+            int word_in_stripe  = _word_in_stripe(i);
+            int tile = _tile(stripe);
+            int stripe_in_tile = _stripe_in_tile(stripe);
+
+            TYPE *ptr =
+                (TYPE*) bsg_remote_ptr(bsg_id_to_x(tile), bsg_id_to_y(tile),
```

Comment:
Use C++ casts!!!! This cast is casting away the volatile-ness of this pointer and will cause incorrect results! If you use `reinterpret_cast<TYPE *>()`, compiler will complain because you have to use `const_cast` to cast away volatile. You want `reinterpret_cast<volatile TYPE *>()`. Dereferencing later down the line will remove volatile on the value.